src/snap/invokeSnap.ts

```diff
@@ -0,0 +1,30 @@
+import { JSONArray, JSONObject, Page } from "puppeteer";
+import { flaskOnly } from "./utils";
+
+export async function invokeSnap<R = unknown>(
+  page: Page,
+  snapId: string,
+  method: string,
+  params: JSONArray | JSONObject = {}
```

Comment:
we should be able to override this with generic

Reply:
@mpetrunic I assume we can do it this way using the generic
```
export type InvokeSnapParams = {
  page: Page;
  snapId: string;
  method: string;
  params: JSONArray | JSONObject;
};

export async function invokeSnap<R, T extends InvokeSnapParams>({
  page,
  snapId,
  params,
  method,
}: T): ReturnType<typeof window.ethereum.request<R>> {
```

But is it really a case for generic as we need some explicit data to pass? I can think of the case when the page is a part of the Playwright framework instead of Puppeteer, but the method calls will be different as well, so may you please suggest the use case of it?

Reply:
we just want snap method params and result as generic, as it would allow us to use snap-provided types to override them and to ensure type safety, page/snapId/method should not be generic.
For example:
```js
import {signSomethingMethod} from "some-snap"

await dappeteer.invokeSnap<
Parameters<typeof signSomethingMethod>,
ReturnType<typeof signSomethingMethod>
>(page, "local:some-snap", "signSomethingMethod", params)

Reply:
@mpetrunic with params it will be tricky to make it generic as we will end up with the same situation I discussed with Bero. We can pass to params only Serialisiable values, so we agreed to JSONArray | JSONObject which are pretty flexible in terms of data we pass. 

```
evaluate<T extends EvaluateFn>(pageFunction: T, ...args: SerializableOrJSHandle[]): Promise<UnwrapPromiseLike<EvaluateFnReturnType<T>>>;
```

Reply:
I agree with that, but we can make `JSONArray | JSONObject` default value for params generic while still allowing people to override for better type safety. cc @BeroBurny

Reply:
@mpetrunic we can try something like this, but it looks hacky 😝
```
export async function invokeSnap<R, P = JSONArray | JSONObject>(
  page: Page,
  snapId: string,
  method: string,
  params: P
): ReturnType<typeof window.ethereum.request<R>> {
  flaskOnly(page);
  return page.evaluate(
    async (opts: { snapId: string; method: string; params: string }) => {
      return window.ethereum.request<R>({
        method: "wallet_invokeSnap",
        params: [
          `${opts.snapId}`,
          {
            method: opts.method,
            params: JSON.parse(opts.params) as object | unknown[],
          },
        ],
      });
    },
    { snapId, method, params: JSON.stringify(params) }
  );
}

Reply:
I'm not sure I understand the problem here. Why would we need to stringify and parse?
params will always be serializable because the snap method can only accept serializable params

Reply:
@mpetrunic params definitely should be serializable but in the case of generic, we can put any value, and then page.evaluate() method is not happy as our generic "P" is not assignable to type 'Serializable'. So, the arguments have for page.evaluate() method strict requirement to be the type SerializableOrJSHandle[]
```
type Serializable = number | string | boolean | null | BigInt | JSONArray | JSONObject;

And the method page.evaluate()  gives us an interface to use generics only for the return method, but not for arguments. Or maybe I missing something 😃

Reply:
Haha thank you. i see now. Then just narrow down generic to be `P extends SerializableOrJSHandle= SerializableOrJSHandle`